File: python/src/rotascale/middleware/openai_compat.py
"""OpenAI-compatible middleware.

Wraps anything exposing `chat.completions.create` — the OpenAI SDK, Azure
OpenAI, Together, Groq, vLLM, Ollama, and every other service that copied the
shape. Duck-typed: this module imports no provider library.

    client = watch_openai(OpenAI())

    with rs.witness("support-agent"):
        client.chat.completions.create(model="gpt-4o", messages=[...])
        # the call is on the trajectory; no other change to the agent
"""

import time
from typing import Any

from rotascale.client import current_trajectory
from rotascale.middleware._common import logger, truncate


class _WatchedCompletions:
    def __init__(self, inner: Any, capture_content: bool, limit: int) -> None:
        self._inner = inner
        self._capture = capture_content
        self._limit = limit

    def create(self, **kwargs: Any) -> Any:
        started = time.perf_counter()
        try:
            response = self._inner.create(**kwargs)
        except Exception as exc:
            # subhadipmitra@: A failed model call is evidence too. "The agent did
            # nothing because the provider was down" is a materially different
            # story from "the agent chose to do nothing", and only one of them
            # is in the record if failures are dropped.
            self._record_failure(kwargs, exc, started)
            raise
        self._record_success(kwargs, response, started)
        return response

    def _step(self, kwargs: dict) -> dict:
        return {
            "provider": "openai-compatible",
            "model_requested": kwargs.get("model"),
            "temperature": kwargs.get("temperature"),
            "tool_choice": kwargs.get("tool_choice"),
        }

    def _record_success(self, kwargs: dict, response: Any, started: float) -> None:
        trajectory = current_trajectory()
        if trajectory is None:
            return
        step = self._step(kwargs)
        step["latency_ms"] = round((time.perf_counter() - started) * 1000, 1)
        step["model_served"] = getattr(response, "model", None)
        step["response_id"] = getattr(response, "id", None)

        usage = getattr(response, "usage", None)
        if usage is not None:
            step["usage"] = {
                "prompt_tokens": getattr(usage, "prompt_tokens", None),
                "completion_tokens": getattr(usage, "completion_tokens", None),
            }

        choices = getattr(response, "choices", None) or []
        if choices:
            first = choices[0]
            step["finish_reason"] = getattr(first, "finish_reason", None)
            message = getattr(first, "message", None)
            calls = getattr(message, "tool_calls", None) if message else None
            if calls:
                # Tool NAMES only. The arguments are the agent's business and
                # land on their own tool_call steps if the agent records them.
                step["tool_calls"] = [
                    getattr(getattr(c, "function", None), "name", None) for c in calls
                ]
            if self._capture and message is not None:
                step["response"] = truncate(getattr(message, "content", None), self._limit)

        if self._capture:
            messages = kwargs.get("messages") or []
            if messages:
                last = messages[-1]
                step["last_message"] = truncate(
                    last.get("content") if isinstance(last, dict) else last, self._limit
                )
        try:
            trajectory.llm_call(**step)
        except Exception:
            logger.warning("rotascale: failed to record llm_call", exc_info=True)

    def _record_failure(self, kwargs: dict, exc: Exception, started: float) -> None:
        trajectory = current_trajectory()
        if trajectory is None:
            return
        step = self._step(kwargs)
        step["latency_ms"] = round((time.perf_counter() - started) * 1000, 1)
        step["error"] = type(exc).__name__
        step["error_message"] = str(exc)[:500]
        try:
            trajectory.llm_call(**step)
        except Exception:
            logger.warning("rotascale: failed to record a failed llm_call", exc_info=True)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._inner, name)


class _WatchedChat:
    def __init__(self, inner: Any, capture_content: bool, limit: int) -> None:
        self._inner = inner
        self.completions = _WatchedCompletions(inner.completions, capture_content, limit)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._inner, name)


class _WatchedClient:
    def __init__(self, inner: Any, capture_content: bool, limit: int) -> None:
        self._inner = inner
        self.chat = _WatchedChat(inner.chat, capture_content, limit)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._inner, name)


def watch_openai(client: Any, *, capture_content: bool = True, content_limit: int = 2000) -> Any:
    """Wrap an OpenAI-compatible client so its calls land on the trajectory.

    `capture_content=False` records shape and metadata only — model, latency,
    tokens, finish reason, tool names — and no prompt or completion text. Some
    deployments cannot put user content in a second store at all, and evidence
    that a customer refuses to enable is worth nothing.
    """
    return _WatchedClient(client, capture_content, content_limit)

File: typescript/src/middleware/openai-compat.ts
/**
 * OpenAI-compatible middleware.
 *
 * Wraps anything exposing `chat.completions.create` — the OpenAI SDK, Azure
 * OpenAI, Together, Groq, vLLM, **Ollama**, and every other service that copied
 * the shape.
 *
 *     const client = watchOpenAI(new OpenAI({ baseURL: "http://localhost:11434/v1" }));
 *     await witness(run, async () => {
 *       await client.chat.completions.create({ model: "llama3.2:3b", messages });
 *       // the call is on the trajectory; no other change to the agent
 *     });
 *
 * subhadipmitra@: DUCK-TYPED, and named `openai-compat` rather than `openai`
 * for the same reason the Python one is: this module imports no provider
 * library and never will. An enterprise running its own models on its own
 * hardware is not a special case to support later — it is the same code path,
 * and calling the file `openai` would imply otherwise.
 *
 * Rotascale governs the ACTION, not the model. Which model produced a request
 * is something the record notes, never something the decision depends on.
 */

import { record, seenServedModel, truncate } from "./common.js";

export interface WatchOptions {
  /**
   * Record prompt and response text.
   *
   * subhadipmitra@: OFF by default. A middleware that captured whole prompts
   * because somebody imported it would put customer data in the evidence store
   * as a side effect of an import, which is not a decision an import may make.
   */
  captureContent?: boolean;
  /** Characters kept per captured field when `captureContent` is on. */
  limit?: number;
}

type Completions = { create(args: Record<string, unknown>): Promise<unknown> };
type ChatClient = { chat: { completions: Completions } };

function get(o: unknown, key: string): unknown {
  return o && typeof o === "object" ? (o as Record<string, unknown>)[key] : undefined;
}

/**
 * Wrap a client so every completion lands on the current trajectory.
 *
 * The returned object delegates everything else to the original, so this is a
 * drop-in: no other change to the agent.
 */
export function watchOpenAI<T extends ChatClient>(
  client: T, options: WatchOptions = {},
): T {
  const capture = options.captureContent ?? false;
  const limit = options.limit ?? 2_000;
  const inner = client.chat.completions;

  const watched: Completions = {
    async create(args: Record<string, unknown>): Promise<unknown> {
      const started = Date.now();
      let response: unknown;
      try {
        response = await inner.create(args);
      } catch (error) {
        // subhadipmitra@: A FAILED call is evidence too. "The agent did nothing
        // because the provider was down" is a materially different story from
        // "the agent chose to do nothing", and only one of them is in the
        // record if failures are dropped.
        await record("llm_call", {
          provider: "openai-compatible",
          model_requested: args.model ?? null,
          latency_ms: Date.now() - started,
          failed: true,
          error: String(error).slice(0, 500),
        });
        throw error;
      }

      const step: Record<string, unknown> = {
        provider: "openai-compatible",
        model_requested: args.model ?? null,
        temperature: args.temperature ?? null,
        tool_choice: args.tool_choice ?? null,
        latency_ms: Date.now() - started,
        model_served: get(response, "model") ?? null,
        response_id: get(response, "id") ?? null,
      };

      // The inventory learns what actually ran, from the runtime that ran it.
      if (seenServedModel(step.model_served as string, "openai-compatible")) {
        step.first_seen_served_model = true;
      }

      const usage = get(response, "usage");
      if (usage) {
        step.usage = {
          prompt_tokens: get(usage, "prompt_tokens") ?? null,
          completion_tokens: get(usage, "completion_tokens") ?? null,
        };
      }

      const choices = get(response, "choices");
      const first = Array.isArray(choices) ? choices[0] : undefined;
      if (first) {
        step.finish_reason = get(first, "finish_reason") ?? null;
        const message = get(first, "message");
        const calls = get(message, "tool_calls");
        if (Array.isArray(calls) && calls.length > 0) {
          // Tool NAMES only. The arguments are the agent's business and land on
          // their own tool_call steps if the agent records them.
          step.tool_calls = calls.map((c) => get(get(c, "function"), "name") ?? null);
        }
        if (capture && message) {
          step.response = truncate(get(message, "content"), limit);
        }
      }

      if (capture && Array.isArray(args.messages) && args.messages.length > 0) {
        const last = args.messages[args.messages.length - 1];
        step.last_message = truncate(get(last, "content") ?? last, limit);
      }

      await record("llm_call", step);
      return response;
    },
  };

  // subhadipmitra@: A PROXY rather than a rebuilt object, so anything else on
  // the client — `embeddings`, `models`, a provider-specific extension —
  // continues to work untouched. Rebuilding would silently drop whatever this
  // file had not heard of, which is every future API.
  return new Proxy(client, {
    get(target, prop, receiver) {
      if (prop === "chat") {
        return new Proxy(target.chat, {
          get(chatTarget, chatProp, chatReceiver) {
            if (chatProp === "completions") return watched;
            return Reflect.get(chatTarget, chatProp, chatReceiver);
          },
        });
      }
      return Reflect.get(target, prop, receiver);
    },
  });
}
